drivers/SmartThings/matter-lock/src/new-matter-lock/init.lua

```diff
@@ -0,0 +1,1670 @@
+  if busy == true then
+    local result = {
+      commandName = cmdName,
+      statusCode = "busy"
+    }
+    local event = capabilities.lockUsers.commandResult(
+      result,
+      {
+        state_change = true,
+        visibility = {displayed = false}
+      }
+    )
+    device:emit_event(event)
+    return
+  end
+
+  -- Save values to field
+  device:set_field(lock_utils.COMMAND_NAME, cmdName, {persist = true})
+  device:set_field(lock_utils.USER_INDEX, ALL_INDEX, {persist = true})
+
+  -- Send command
+  local ep = device:component_to_endpoint(command.component)
+  device:send(DoorLock.server.commands.ClearUser(device, ep, ALL_INDEX))
+end
+
+-------------------------
+-- Clear User Response --
+-------------------------
+local function clear_user_response_handler(driver, device, ib, response)
+  -- Get result
+  local cmdName = device:get_field(lock_utils.COMMAND_NAME)
+  local userIdx = device:get_field(lock_utils.USER_INDEX)
+  local status = "success"
+  if ib.status == DoorLock.types.DlStatus.FAILURE then
+    status = "failure"
+  elseif ib.status == DoorLock.types.DlStatus.INVALID_FIELD then
+    status = "invalidCommand"
+  end
+
+  -- Delete User and Credential from table
+  if status == "success" then
+    delete_user_from_table(device, userIdx)
+    delete_credential_from_table_as_user(device, userIdx)
+  else
+    device.log.warn(string.format("Failed to clear user: %s", status))
+  end
+
+  -- Update commandResult
+  local result = {
+    commandName = cmdName,
+    userIndex = userIdx,
+    statusCode = status
+  }
+  local event = capabilities.lockUsers.commandResult(
+    result,
+    {
+
+      visibility = {displayed = false}
+    })
+  device:emit_event(event)
+  device:set_field(lock_utils.BUSY_STATE, false, {persist = true})
+end
+
+--------------------
+-- Add Credential --
+--------------------
+local function handle_add_credential(driver, device, command)
+  -- Get parameters
+  local cmdName = "addCredential"
+  local userIdx = command.args.userIndex
+  if userIdx == 0 then
+    userIdx = nil
+  end
+  local userType = command.args.userType
+  local userTypeMatter = DoorLock.types.UserTypeEnum.UNRESTRICTED_USER
+  if userType == "guest" then
+    userTypeMatter = DoorLock.types.UserTypeEnum.SCHEDULE_RESTRICTED_USER
+  end
+  local credential = {
+    credential_type = DoorLock.types.CredentialTypeEnum.PIN,
+    credential_index = INITIAL_COTA_INDEX
+  }
+  local credData = command.args.credentialData
+
+  -- Check busy state
+  local busy = check_busy_state(device)
+  if busy == true then
+    local result = {
+      commandName = cmdName,
+      statusCode = "busy"
+    }
+    local event = capabilities.lockCredentials.commandResult(
+      result,
+      {
+        state_change = true,
+        visibility = {displayed = false}
+      }
+    )
+    device:emit_event(event)
+    return
+  end
+
+  -- Save values to field
+  device:set_field(lock_utils.COMMAND_NAME, cmdName, {persist = true})
```

Comment:
@HunsupJung I understand how you are using these fields to save information to be used in a different function, but could you help me understand why we need to do it this way? Do the commands themselves not have the relevant information that we can use in the handler? It seems this is the case for things like the ClearUser command, but do we really need to save all of this information or can be pull some of it from the command args and ib elements?

Also, it seems like there could be issues if two commands that rely on a field like `USER_INDEX` are processed at the same time. For example, if one command sets `USER_INDEX` to 1 and then before we handle that response, another command sets `USER_INDEX` to 2 , the original value would be overwritten and could cause problems in handling the first command that expected `USER_INDEX` to be 1. Are there cases where this could happen?

Reply:
> Also, it seems like there could be issues if two commands that rely on a field like USER_INDEX are processed at the same time. For example, if one command sets USER_INDEX to 1 and then before we handle that response, another command sets USER_INDEX to 2 , the original value would be overwritten and could cause problems in handling the first command that expected USER_INDEX to be 1. Are there cases where this could happen?

Therefore, I used busy_state to proceed with only one command at a time.